Library/Homebrew/cmd/uninstall.rb

```diff
@@ -26,6 +53,36 @@ def uninstall
         name = Formulary.canonical_name(name)
         rack = HOMEBREW_CELLAR/name
 
+        # Cellar/oldname can be a symlink to Cellar/newname so in order not to
+        # get newname's keg as oldname's keg we check if it's a symlink.
+        if rack.symlink?
+          raise "#{rack.to_s} is a symlink"
+        end
+
+        # Since we delete every keg, opt/oldname and Cellar/oldname symlinks
+        # must be removed.
+        begin
+          formula = Formulary.from_rack(rack)
```

Comment:
Again, I think we can avoid load formula completely. So we can have a more robust user experience.